StitchCore/StitchCore/Source/Core/StitchClient.swift

```diff
@@ -268,27 +276,41 @@ public class StitchClient: StitchClientType {
     public func login(withProvider provider: AuthProvider) -> Promise<UserId> {
         self.authProvider = provider
 
-        if isAuthenticated, let auth = auth {
-            printLog(.info, text: "Already logged in, using cached token.")
-            return Promise.init(value: auth.userId)
+        func doLoginRequest() -> Promise<UserId> {
+            return httpClient.doRequest {
+                $0.method = .post
+                $0.endpoint = self.routes.authProvidersLoginRoute(provider: provider.type)
+                $0.isAuthenticatedRequest = false
+                try $0.encode(withData: self.getAuthRequest(provider: provider))
+                }.flatMap { [weak self] any in
+                    guard let strongSelf = self else { throw StitchError.clientReleased }
+                    let authInfo = try JSONDecoder().decode(AuthInfo.self,
+                                                            from: JSONSerialization.data(withJSONObject: any))
+                    strongSelf.httpClient.authInfo = authInfo
+                    strongSelf._auth = Auth(stitchClient: strongSelf,
+                                            stitchHttpClient: strongSelf.httpClient,
+                                            userId: authInfo.userId)
+                    strongSelf.onLogin()
+                    return authInfo.userId
+                }
         }
 
-        return httpClient.doRequest {
-            $0.method = .post
-            $0.endpoint = self.routes.authProvidersLoginRoute(provider: provider.type)
-            $0.isAuthenticatedRequest = false
-            try $0.encode(withData: self.getAuthRequest(provider: provider))
-        }.flatMap { [weak self] any in
-            guard let strongSelf = self else { throw StitchError.clientReleased }
-            let authInfo = try JSONDecoder().decode(AuthInfo.self,
-                                                    from: JSONSerialization.data(withJSONObject: any))
-            strongSelf.httpClient.authInfo = authInfo
-            strongSelf._auth = Auth(stitchClient: strongSelf,
-                                    stitchHttpClient: strongSelf.httpClient,
-                                    userId: authInfo.userId)
-            strongSelf.onLogin()
-            return authInfo.userId
+        if isAuthenticated, let auth = auth {
+            if provider.type == AuthProviderTypes.anonymous.rawValue &&
+                loggedInProviderType == AuthProviderTypes.anonymous.rawValue {
+                printLog(.info, text: "Already logged in as anonymous user, using cached token.")
+                return Promise.init(value: auth.userId)
+            }
+
+            // Using a different provider, log outand then perform login.
+            printLog(.info, text: "Already logged in, logging out of existing session.")
+            return self.logout().then {
```

Comment:
[minor] If logout fails, we still want to log in. See how the JS SDK does this in the logout method by handling both resolve and reject cases.

Reply:
Done.

That being said, I'm not sure this is the behavior currently in the JS SDK, because if you look at this line: https://github.com/mongodb/stitch-js-sdk/blob/master/src/client.js#L171

`authenticateFn` is only being called on the resolve case of the `logout` promise. The `logout` method handling both resolve and reject in the JS SDK is for clearing auth in storage, not for authenticating.